src/infrastructure/selectors/XmlSelectorAdapter.js

```diff
@@ -0,0 +1,44 @@
+const xpath = require('xpath');
+const { XMLSerializer, DOMParser } = require('xmldom');
+const DataSelectorPort = require('../../domain/common/port/DataSelectorPort');
+
+class XmlSelectorAdapter extends DataSelectorPort {
+  constructor() {
+    super();
+    this.domParser = new DOMParser();
+    this.xmlSerializer = new XMLSerializer();
+  }
+
+  canHandle(contentType) {
+    return ['xml', 'html'].includes(contentType);
+  }
+
+  select(data, path) {
+    const doc = this.domParser.parseFromString(data);
+    this.xmlSerializer.serializeToString(doc); // this throw an error when doc is invalid xml
+
+    if (path === null) {
+      return doc.toString();
+    }
+
+    const selected = xpath.select(path, doc);
+
+    if (selected instanceof Array && selected.length === 0) {
+      return null;
+    }
+
+    if (selected instanceof Array && selected.length > 1) {
+      const parsedResults = selected.map((el) => {
```

Comment:
you can use destructor here `map(({ data }) => {

Reply:
Not really, he wouldn't be able to do a `return el;` at the end then, right?

Reply:
of course! my bad